### ISSUE TITLE
ES3 reserved words

### ISSUE THREAD

**Issue by zloirock** [eslint/eslint#15017]
The issue: https://github.com/zloirock/core-js/issues/980

In the code used `char` as a variable name. [The ESLint config contains `ecmaVersion: 3`](https://github.com/zloirock/core-js/blob/a3f2dc01bc045c117da74779708e1acb0f0d38e5/.eslintrc.js#L993), till this week here was used the default ESLint parser. `char` is a ES3 reserved word:

<img width="565" alt="Снимок экрана 2021-09-01 в 20 46 37" src="https://user-images.githubusercontent.com/2213682/131682898-2b2acb24-f476-4aa8-84a5-d29d34001d41.png">

ESLint parsed it without any errors or warnings. However, this code might not work in some ES3 engines.

**Comment by nzakas** [comment]
Please update your issue to use our [bug report template](https://github.com/eslint/eslint/blob/master/.github/ISSUE_TEMPLATE/BUG_REPORT.md) and we will take a look. Thanks.

**Comment by aladdin-add** [comment]
I was able to repro, and it seems an issue in acorn.

**Comment by mdjermanovic** [comment]
Acorn has an option to modify this behavior:

https://github.com/acornjs/acorn/tree/master/acorn#interface

> - **allowReserved**: If `false`, using a reserved word will generate
>   an error. Defaults to `true` for `ecmaVersion` 3, `false` for higher
>   versions. When given the value `"never"`, reserved words and
>   keywords can also not be used as property names (as in Internet
>   Explorer's old parser).

Consequently, `espree` (ESLint's default parser, which uses `acorn`) allows reserved words when `ecmaVersion: 3` is specified, since we are not setting this option so the defaults apply.

I'm not sure what was the reasoning for these defaults in `acorn`. With `allowReserved: false`, it does throw a parsing error ([astexplorer.net snippet](https://astexplorer.net/#/gist/39b93153128ee6cf2230a86d6de45ed1/95e27c601f3b484ca961f4bfb6c834a45eccf535)). 

We could consider setting `allowReserved: false` [here](https://github.com/eslint/espree/blob/63bd0bc46adfbcd3a71d4cc222aa3923b76ebcf2/lib/espree.js#L66), but that feels like a breaking change. Thoughts?

**Comment by zloirock** [comment]
As an option, instead of parsing error, it can be a new rule that will forbid the usage of ES3 reserved words as variables names.

**Comment by mdjermanovic** [comment]
[id-denylist](https://eslint.org/docs/rules/id-denylist) also provides similar functionality, though with some exceptions (e.g., property names are allowed).

**Comment by zloirock** [comment]
@mdjermanovic I know, but I don't think that it's fine to recommend adding the ES3 reserved words list to those rules for all who need ES3 support.

**Comment by aladdin-add** [comment]
the easiest way is to use the `allowReserved` option:

```js
"parserOptions": { "ecmaVersion": 3, "allowReserved": false}
```

I also agreed this can be the default when ecmaVersion is 3.

**Comment by mdjermanovic** [comment]
> As an option, instead of parsing error, it can be a new rule that will forbid the usage of ES3 reserved words as variables names.
>
> @mdjermanovic I know, but I don't think that it's fine to recommend adding the ES3 reserved words list to those rules for all who need ES3 support.

I agree this isn't an ideal solution, but it's a currently available workaround. We don't have core rules to check syntax, that rule seems more like a candidate for [eslint-plugin-es](https://github.com/mysticatea/eslint-plugin-es).

> ```js
> "parser": "espree",
> "parserOptions": { "ecmaVersion": 3, "allowReserved": false}
> ```

This would solve the problem in a non-breaking way, but we aim to avoid exposing Acorn options for various reasons.

I think that `espree` should just always pass `"allowReserved": false` to `acorn`, but that would have to wait for espree/eslint v9.

**Comment by zloirock** [comment]
> eslint-plugin-es

This plugin seems [dead](https://github.com/mysticatea/eslint-plugin-es/issues/72). Anyway, this issue looks like an ESLint core problem.

> that would have to wait for espree/eslint v9

Why not add it to v8 while it's still in beta?

**Comment by aladdin-add** [comment]
filed the issue in acorn: https://github.com/acornjs/acorn/issues/1062 to see if they want to change the default.

**Comment by mdjermanovic** [comment]
> Why not add it to v8 while it's still in beta?

[espree v8.0.0](https://www.npmjs.com/package/espree) has been already released, and the changes in eslint v8.0.0 are already [announced](https://eslint.org/blog/2021/06/whats-coming-in-eslint-8.0.0).

**Comment by nzakas** [comment]
Because there’s agreement that this is a bug, I’m marking as accepted. We just need to determine how to proceed.

I’m guessing we don’t have a lot of ES3 usage right now, and ES3 code using reserved words will throw a runtime error, so even though this is a breaking change, I think the impact to end users would be minimal. So, we may want to evaluate this a bit differently than other breaking changes.

**Comment by mdjermanovic** [comment]
We could fix this in Espree as a breaking change, release Espree v9.0.0, and then use Espree 9 in ESLint v8.0.0.

**Comment by ljharb** [comment]
All of my packages are authored in ES3, and none of them use `char` as a variable name, so I'd love to see this fix come in, even as a patch.

**Comment by mdjermanovic** [comment]
Should we actually set `allowReserved: "never"` for ES3, to also disallow `obj.char` and `var obj = { char: 1 }`?

By https://www-archive.mozilla.org/js/language/E262-3.pdf (if that's the right document?):

> _Identifier_ ::
> &nbsp;&nbsp;&nbsp;&nbsp;  _IdentifierName_ but not _ReservedWord_

> _MemberExpression_ :
> &nbsp;&nbsp;&nbsp;&nbsp; ...
> &nbsp;&nbsp;&nbsp;&nbsp; _MemberExpression_ . ___Identifier___

> _ObjectLiteral_ :
> &nbsp;&nbsp;&nbsp;&nbsp; { }
> &nbsp;&nbsp;&nbsp;&nbsp; { _PropertyNameAndValueList_ }
> _PropertyNameAndValueList_ :
> &nbsp;&nbsp;&nbsp;&nbsp; _PropertyName_ : _AssignmentExpression_
> &nbsp;&nbsp;&nbsp;&nbsp; _PropertyNameAndValueList_ , _PropertyName_ : _AssignmentExpression_
> PropertyName :
> &nbsp;&nbsp;&nbsp;&nbsp; ___Identifier___
> &nbsp;&nbsp;&nbsp;&nbsp; _StringLiteral_
> &nbsp;&nbsp;&nbsp;&nbsp; _NumericLiteral_

**Comment by nzakas** [comment]
> We could fix this in Espree as a breaking change, release Espree v9.0.0, and then use Espree 9 in ESLint v8.0.0.

That’s what I was thinking, too. 

> Should we actually set allowReserved: "never" for ES3, to also disallow obj.char and var obj = { char: 1 }?

Yes. Reserved words were completely illegal in ES3.

**Comment by nzakas** [comment]
Note from marijn about Acorn's default:

> Most browsers in the ES3 era allowed reserved words to be used in scripts, hence this default. It is not going to change for backwards-compatibility reasons.

**Comment by nzakas** [comment]
Opened a PR: https://github.com/eslint/espree/pull/513

**Comment by mdjermanovic** [comment]
In yesterday's TSC meeting, we decided to accept this change for ESLint v8.0.0.

The change will be made directly in Espree. We'll release Espree v9.0.0 and use Espree v9 in ESLint v8.0.0.

Reserved words will also be disallowed in property names, such as `obj.char` or `var obj = { char: 1 }`. The new behavior is spec-compliant, and it isn't configurable.

This change only affects ESLint users with the default parser and `"parserOptions": { "ecmaVersion": 3 }`.
